mypy/checker.py

```diff
@@ -1452,6 +1460,55 @@ def check_simple_assignment(self, lvalue_type: Type, rvalue: Expression,
                                    '{} has type'.format(lvalue_name))
             return rvalue_type
 
+    def check_descriptor_assignment(self, instance_type: Type, descriptor_type: Type,
+                                    rvalue: Expression, context: Context) -> Tuple[Type, bool]:
+        """
+        Type check descriptor assignment.
+
+        Return the inferred rvalue_type and whether to infer anything about the lvalue type
+        """
+        # Descriptors don't participate in class-attribute access
+        if isinstance(instance_type, FunctionLike) and instance_type.is_type_obj():
+            rvalue_type = self.check_simple_assignment(descriptor_type, rvalue, context)
+            return rvalue_type, True
+
+        if not isinstance(descriptor_type, Instance):
+            rvalue_type = self.check_simple_assignment(descriptor_type, rvalue, context)
+            return rvalue_type, True
+
+        if not descriptor_type.type.has_readable_member('__set__'):
+            if descriptor_type.type.has_readable_member('__get__'):
+                descriptor_type = self.expr_checker.analyze_descriptor_access(
+                    instance_type, descriptor_type, context
+                )
+            rvalue_type = self.check_simple_assignment(descriptor_type, rvalue, context)
+            return rvalue_type, True
+
+        dunder_set = descriptor_type.type.get_method('__set__')
+
+        if dunder_set is None:
+            self.msg.fail("{}.__set__ is not callable".format(descriptor_type), context)
+            return AnyType(), False
+
+        function = function_type(dunder_set, self.named_type('builtins.function'))
+        bound_method = bind_self(function, descriptor_type)
+        typ = map_instance_to_supertype(descriptor_type, dunder_set.info)
+        dunder_set_type = expand_type_by_instance(bound_method, typ)
+
+        _, inferred_dunder_set_type = self.expr_checker.check_call(
+            dunder_set_type, [TempNode(instance_type), rvalue],
+            [nodes.ARG_POS, nodes.ARG_POS], context
+        )
+        if not isinstance(inferred_dunder_set_type, CallableType):
+            self.fail("__set__ was not callable", context)
```

Comment:
Fixed the comment. I have tests for when `__set__` is a string (which I felt stood in for any of the non-callable types. The `AnyType()` being returned is because I thought that was the standard return type when an type inference error had occurred.